dbms/src/Flash/Mpp/ExchangeReceiver.cpp

```diff
@@ -740,56 +725,58 @@ ExchangeReceiverResult ExchangeReceiverBase<RPCContext>::nextResult(std::queue<B
-        {
-            return ExchangeReceiverResult::newEOF(name);
-        }
+        return state != ExchangeReceiverState::NORMAL
+            ? ExchangeReceiverResult::newError(0, name, constructStatusString(state, err_msg))
+            : ExchangeReceiverResult::newEOF(name); /// live_connections == 0, msg_channel is finished, and state is NORMAL, that is the end.
     }
-    assert(recv_msg != nullptr);
-    ExchangeReceiverResult result;
-    if (recv_msg->error_ptr != nullptr)
+    else
     {
-        result = ExchangeReceiverResult::newError(recv_msg->source_index, recv_msg->req_info, recv_msg->error_ptr->msg());
+        assert(recv_msg != nullptr);
+        if (unlikely(recv_msg->error_ptr != nullptr))
+            return ExchangeReceiverResult::newError(recv_msg->source_index, recv_msg->req_info, recv_msg->error_ptr->msg());
+        return toDecodeResult(block_queue, header, recv_msg);
     }
-    else
+}
+
+template <typename RPCContext>
+ExchangeReceiverResult ExchangeReceiverBase<RPCContext>::toDecodeResult(
+    std::queue<Block> & block_queue,
+    const Block & header,
+    const std::shared_ptr<ReceivedMessage> & recv_msg)
+{
+    assert(recv_msg != nullptr);
+    if (recv_msg->resp_ptr != nullptr) /// the data of the last packet is serialized from tipb::SelectResponse including execution summaries.
     {
-        if (recv_msg->resp_ptr != nullptr) /// the data of the last packet is serialized from tipb::SelectResponse including execution summaries.
+        auto select_resp = std::make_shared<tipb::SelectResponse>();
+        if (unlikely(!select_resp->ParseFromString(*(recv_msg->resp_ptr))))
+        {
+            return ExchangeReceiverResult::newError(recv_msg->source_index, recv_msg->req_info, "decode error");
+        }
+        else
         {
-            auto select_resp = std::make_shared<tipb::SelectResponse>();
-            if (!select_resp->ParseFromString(*(recv_msg->resp_ptr)))
+            auto result = ExchangeReceiverResult::newOk(select_resp, recv_msg->source_index, recv_msg->req_info);
+            /// If mocking TiFlash as TiDB, here should decode chunks from select_resp.
+            if (unlikely(!result.resp->chunks().empty()))
             {
-                result = ExchangeReceiverResult::newError(recv_msg->source_index, recv_msg->req_info, "decode error");
+                assert(recv_msg->chunks.empty());
+                // Fine grained shuffle should only be enabled when sending data to TiFlash node.
+                // So all data should be encoded into MPPDataPacket.chunks.
+                RUNTIME_CHECK_MSG(!enableFineGrainedShuffle(fine_grained_shuffle_stream_count), "Data should not be encoded into tipb::SelectResponse.chunks when fine grained shuffle is enabled");
+                result.decode_detail = CoprocessorReader::decodeChunks(select_resp, block_queue, header, schema);
             }
-            else
+            else if (!recv_msg->chunks.empty())
```

Comment:
Is the  check "(!recv_msg->chunks.empty())" necessary?

Reply:
For the last packet, chunks may be empty.